packages/diagram/lib/diagram.dart

```diff
@@ -0,0 +1,442 @@
+        new Quaternion.identity(), // rotation
+        new Vector3(scale, scale, 1.0) // scale
+        );
+    return matrix;
+  }
+
+  final Matrix4 _paintMatrix;
+  final Matrix4 _hitTestMatrix;
+
+  @override
+  Matrix4 toMatrix() => _paintMatrix.clone();
+
+  /// Provides the transformation matrix that converts coordinates in the test
+  /// coordinate space to coordinates in logical pixels on the real display.
+  ///
+  /// This is essentially the same as [toMatrix] but ignoring the device pixel
+  /// ratio.
+  ///
+  /// This is useful because pointers are described in logical pixels, as
+  /// opposed to graphics which are expressed in physical pixels.
+  Matrix4 toHitTestMatrix() => _hitTestMatrix.clone();
+
+  @override
+  String toString() => 'TestViewConfiguration';
+}
+
+// Provides a binding different from the regular Flutter binding so that
+// diagrams can control their timeline and physical device size.
+class _DiagramFlutterBinding extends BindingBase
+    with
+        GestureBinding, //
+        ServicesBinding,
+        SchedulerBinding,
+        PaintingBinding,
+        RendererBinding,
+        WidgetsBinding {
+  @override
+  void initInstances() {
+    super.initInstances();
+    // Short circuit the begin frame to use our timestamp instead of the "real" one.
+    // Re-use the previous onBeginFrame so that all of the hot reload, etc. works.
+    dynamic oldBeginFrame = ui.window.onBeginFrame;
+    ui.window.onBeginFrame = (Duration timestamp) {
+      oldBeginFrame(_timestamp);
+    };
+  }
+
+  Duration _timestamp = Duration.zero;
+  GlobalKey _boundaryKey = new GlobalKey();
+
+  /// Determines the ratio between physical units and logical units.
+  ///
+  /// The [pixelRatio] describes the scale between the logical pixels and the
+  /// size of the output image. It is independent of the
+  /// [window.devicePixelRatio] for the device, so specifying 1.0 (the default)
+  /// will give you a 1:1 mapping between logical pixels and the output pixels
+  /// in the image. The output image is not anti-aliased, so you may wish to
+  /// use a larger device pixel ratio to obtain an image large enough to
+  /// down sample.
+  ///
+  /// Defaults to 1.0.
+  double get pixelRatio => _pixelRatio;
+  double _pixelRatio = 1.0;
+  set pixelRatio(double pixelRatio) {
+    _pixelRatio = pixelRatio;
+    handleMetricsChanged();
+  }
+
+  /// Determines the dimensions of the virtual screen that the diagram will
+  /// be drawn on, in logical units.
+  Size get screenDimensions => _screenDimensions;
+  Size _screenDimensions = _kDefaultDiagramViewportSize;
+  set screenDimensions(Size screenDimensions) {
+    _screenDimensions = screenDimensions;
+    handleMetricsChanged();
+  }
+
+  /// Returns the singleton for the binding, after making sure it is
+  /// initialized.
+  static WidgetsBinding instance() {
+    if (WidgetsBinding.instance == null) {
+      new _DiagramFlutterBinding();
+    }
+    return WidgetsBinding.instance;
+  }
```

Comment:
see https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/scheduler/binding.dart#L194 for the idiomatic way to do this.